Feature Project/Feature Plugin for Plugin Dependencies

afragen opened the pull request:
<!--
Hi there! Thanks for contributing to WordPress!

Pull Requests in this GitHub repository **must** be linked to a ticket in the WordPress Core Trac instance (https://core.trac.wordpress.org), and are only used for code review. **No pull requests will be merged on GitHub.**

See the WordPress Handbook page on using PRs for Code Review more information: https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/

If this is your first time contributing, you may also find reviewing these guides first to be helpful:
- FAQs for New Contributors: https://make.wordpress.org/core/handbook/tutorials/faq-for-new-contributors/
- Contributing with Code Guide: https://make.wordpress.org/core/handbook/contribute/
- WordPress Coding Standards: https://make.wordpress.org/core/handbook/best-practices/coding-standards/
- Inline Documentation Standards: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/
- Browser Support Policies: https://make.wordpress.org/core/handbook/best-practices/browser-support/
- Proper spelling and grammar related best practices: https://make.wordpress.org/core/handbook/best-practices/spelling/
-->

A [Feature Project](https://make.wordpress.org/core/2022/02/24/feature-project-plugin-dependencies/) for Plugin Dependencies has been created. Along with this a [Feature Plugin](https://wordpress.org/plugins/wp-plugin-dependencies/) has also been created.

Testing either via this PR or the feature plugin will show similar/identical results.

Trac ticket: https://core.trac.wordpress.org/ticket/22316

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

afragen committed:
update test

afragen committed:
fix tests

afragen committed:
add slugs with symbols

afragen committed:
more tests

afragen committed:
Merge branch 'WordPress:trunk' into dependencies-tab

afragen committed:
sync with feature plugin

afragen committed:
unused

afragen committed:
Revert "Delete package-lock.json"

This reverts commit 304c704868e46c85e37608d94cabca9ace2b422f.

afragen committed:
Merge branch 'trunk' into dependencies-tab

afragen commented:
Somehow I need to remove `package-lock.json` from this PR.

afragen committed:
hopefully unf*ck package-lock.json

afragen committed:
fix WPCS

afragen committed:
fix multisite compatibility

afragen committed:
contributors is an array

afragen committed:
initialize with hook and static callback

afragen committed:
add filter for switching non-premium slug to premium slug

afragen committed:
more error checking

afragen committed:
circular dependencies, update to plugin

afragen committed:
Merge branch 'trunk' into dependencies-tab

afragen committed:
update @since

afragen committed:
fix for PHP coding standards

afragen committed:
update since

afragen committed:
Merge branch 'trunk' into dependencies-tab

afragen committed:
Merge branch 'trunk' into dependencies-tab

afragen committed:
cleanup docblocks

afragen committed:
initialize on loading class

afragen committed:
update for a11y

afragen committed:
Merge branch 'trunk' into dependencies-tab

afragen committed:
move filter upwards, more versatile

afragen committed:
update action links

afragen committed:
update regex, allow underscores, not symbols

afragen committed:
does need to run during AJAX

afragen committed:
update generic plugin card

afragen committed:
update regex to conform to plugin directory standards

afragen committed:
update generic card short description

afragen committed:
make 'Cannot Install' button when package is empty

afragen committed:
update conditional for button

afragen committed:
forgot new tests

afragen committed:
add tests, but skip until Plugin Dependencies part 2

afragen committed:
don't skip, just remove until needed

afragen committed:
new line before return

afragen committed:
return generic plugin card data when needed

don't hide

afragen committed:
fix conditional logic, more data for empty plugin card

afragen committed:
add more data to generic plugin card

afragen committed:
wrap in p tag, spacing

afragen committed:
update for WP-CLI activate

afragen committed:
no need to start on hook

afragen committed:
Merge branch 'trunk' into dependencies-tab

afragen committed:
add function to create plugin install buttons and style

afragen commented:
After #design has been completed the PR must be updated.

afragen committed:
update since

afragen committed:
bring inline with plugin, simpler plugin card

only with More details link, can install dependency modal

afragen committed:
update button link to show Cannot Install as appropriate

afragen committed:
update More details link

afragen committed:
updated css

afragen committed:
updated to remove Dependencies tab and more of Ozz's feedback

afragen committed:
wpcs fix

afragen committed:
remove as part of Dependencies tab

afragen committed:
add support for single file plugins

afragen committed:
Revert "add support for single file plugins"

This reverts commit 4fc255c3905b990b9c0b6ed52e75f4a5c52b8421.

afragen committed:
move to common.css

afragen committed:
remove from class

afragen committed:
move to src/js/_enqueues/admin

afragen committed:
hopefully passes pagenow to function

afragen committed:
Merge branch 'trunk' into dependencies-tab

afragen committed:
deactivate Install Now button for empty package

afragen committed:
rename function

afragen committed:
make some callbacks static

afragen committed:
lint

afragen committed:
add filter to remove JS method of adding 'Requires:' to plugin row meta

afragen committed:
add 'post_plugin_row_meta' filter

allows adding messaging after the plugin row meta and removes previous JS

afragen committed:
check variable isset

afragen committed:
add filter to hide plugin row checkbox

afragen committed:
update to use filter to hide plugin row checkbox

afragen committed:
must be static

afragen committed:
make class static and move to WPINC

afragen committed:
load class

afragen committed:
load WP_Plugin_Dependencies filters

afragen committed:
rearrange

afragen committed:
initialize() not static and run after require

costdev commented on src/wp-admin/includes/class-wp-plugins-list-table.php:
```diff
@@ -1149,6 +1161,17 @@ public function single_row( $item ) {
 
 					echo '</div>';
 
+					/**
+					 * Add text at end of plugin row meta.
+					 *
+					 * @since 6.4.0
+					 *
+					 * @param string null         Text to add after plugin row meta.
+					 * @param string $plugin_file Refer to 'plugin_row_meta' filter.
+					 * @param array  $plguin_data Refer to 'plugin_row_meta' filter.
+					 */
+					echo apply_filters( 'post_plugin_row_meta', null, $plugin_file, $plugin_data );
```
```suggestion
					/**
					 * Filters text to add after plugin row meta.
					 *
					 * @since 6.4.0
					 *
					 * @param string $after       Text to add after plugin row meta. Default empty string.
					 * @param string $plugin_file Refer to 'plugin_row_meta' filter.
					 * @param array  $plugin_data Refer to 'plugin_row_meta' filter.
					 */
					$after_row_meta = apply_filters( 'after_plugin_row_meta', '', $plugin_file, $plugin_data );

					if ( '' !== $after_row_meta ) {
						echo esc_html( $after_row_meta );
					}
```
Suggested changes:
- Change the summary to start with "Filters..." and to align language with the parameter description.
- Only use `string` for the type, with an empty string being the default value.
- Add missing parameter name and correct a typo in the docblock.
- Rename `post_` to `after_` to avoid confusion with post-related hooks.
- Check for an empty string. Otherwise, escape and output.

costdev replied:
Actually, this may be better as an Action hook instead, with the escaping happening in the callback. What do you think?

afragen replied:
That would work

costdev replied:
```suggestion
					/**
					 * Fires after plugin row meta.
					 *
					 * @since 6.4.0
					 *
					 * @param string $plugin_file Refer to {@see 'plugin_row_meta'} filter.
					 * @param array  $plugin_data Refer to {@see 'plugin_row_meta'} filter.
					 */
					do_action( 'after_plugin_row_meta', $plugin_file, $plugin_data );
```

afragen committed:
use require_once

afragen committed:
remove tests for function not in class

afragen committed:
Update src/wp-admin/includes/class-wp-plugins-list-table.php

change to action hook

Co-authored-by: Colin Stewart <79332690+costdev@users.noreply.github.com>

costdev commented on src/wp-admin/includes/class-wp-plugins-list-table.php:
```diff
@@ -1149,6 +1161,16 @@ public function single_row( $item ) {
+					 * @since 6.4.0
+					 *
+					 * @param string $plugin_file Refer to {@see 'plugin_row_meta'} filter.
+					 * @param array  $plguin_data Refer to {@see 'plugin_row_meta'} filter.
```
```suggestion
					 * @param array  $plugin_data Refer to {@see 'plugin_row_meta'} filter.
```
Bah, I missed the typo! :joy:

afragen committed:
Merge remote-tracking branch 'afragen/dependencies-tab' into dependencies-tab

afragen committed:
fix typo

afragen committed:
change to action hook

afragen committed:
back to static

afragen committed:
update as escaped at output

afragen committed:
update CSS and move to list-tables.css

afragen committed:
updated for new CSS

afragen committed:
missed this bit of CSS for plugins row

afragen committed:
don't need parens for echo

afragen committed:
declare variables with values, remove constructor

afragen committed:
use is_array instead of isset

afragen committed:
refactor

afragen committed:
missed this for plugin card iframe

afragen committed:
move hooks to default-filters.php so they load earlier

Necessary if extending WP_Plugin_Dependencies

afragen committed:
Merge branch 'trunk' into dependencies-tab

afragen committed:
run load in is_admin

afragen committed:
correctly merge updates.js

afragen committed:
update plugin card

afragen committed:
make identical

afragen committed:
Merge branch 'trunk' into dependencies-tab

afragen committed:
update for refactor

afragen committed:
overly aggressive changing 'name' to 'Name'

afragen committed:
use '! empty' as might be false

afragen committed:
updated tests

afragen committed:
make unique set

afragen committed:
lint

afragen committed:
Revert "lint"

This reverts commit 5e1264251e049ee2a37e4ecaf87ee24d3507743f.

afragen committed:
only my tests

afragen committed:
update logic

afragen committed:
update tests

afragen committed:
lint

afragen committed:
rearrange the deck chairs

afragen committed:
lint something

afragen committed:
add docblock

afragen committed:
Revert "move testing directory"

This reverts commit ce3d9068dd4019ea689fa2b8db23d176585d1529.

afragen committed:
move testing directory

afragen committed:
change directory name

afragen committed:
update phpcs rules to skip these UnitTestCase files

afragen committed:
update tests

Commit:
Refactor

afragen committed:
Merge pull request #4 from costdev/plugin-dependencies-refactor

Refactor.

costdev commented on src/wp-admin/includes/class-wp-plugin-install-list-table.php:
```diff
@@ -719,4 +721,87 @@ public function display_rows() {
+	 *
+	 * @since 6.4.0
+	 *
+	 * @param array  $plugin_data An array of plugin data. See {@see plugins_api()}
+	 *                            for the list of possible values.
```
```suggestion
	 * @param string $name The plugin's name.
	 * @param string $slug The plugin's slug.
```

costdev commented on src/wp-admin/includes/class-wp-plugins-list-table.php:
```diff
@@ -1399,4 +1413,107 @@ public function single_row( $item ) {
+			|| $name === $slug
+			|| $name !== $dependency_data['name']
+			|| empty( $dependency_data['version'] )
+			|| $name !== $dependency_data['name']
```
```suggestion
```

afragen committed:
Update src/wp-admin/includes/class-wp-plugin-install-list-table.php

Co-authored-by: Colin Stewart <79332690+costdev@users.noreply.github.com>

afragen committed:
Update src/wp-admin/includes/class-wp-plugins-list-table.php

Co-authored-by: Colin Stewart <79332690+costdev@users.noreply.github.com>

afragen committed:
remove from phpcs.xml.dist

afragen committed:
Merge remote-tracking branch 'afragen/dependencies-tab' into dependencies-tab

costdev commented on src/wp-admin/includes/class-wp-plugins-list-table.php:
```diff
@@ -808,14 +811,19 @@ public function single_row( $item ) {
+									esc_attr( $plugin_id_attr ),
+									/* translators: %s: Plugin name. */
+									esc_attr( sprintf( _x( 'Network Activate %s', 'plugin' ), $plugin_data['Name'] ) ),
+									__( 'Network Activate' ),
```
```suggestion
									__( 'Network Activate' )
```

afragen committed:
Update src/wp-admin/includes/class-wp-plugins-list-table.php

Co-authored-by: Colin Stewart <79332690+costdev@users.noreply.github.com>

costdev commented on src/wp-includes/class-wp-plugin-dependencies.php:
```diff
@@ -0,0 +1,884 @@
+	 * @return bool Whether the plugin has plugins that depend on it.
+	 */
+	public static function has_dependents( $plugin_file ) {
+		$dependency_paths = self::get_dependency_filepaths();
+		return in_array( $plugin_file, $dependency_paths, true );
```
No need to run the logic of `::get_dependency_filepaths()`. Let's just slugify it and check it against the dependency slugs.
```suggestion
		$slug = str_contains( $plugin_file, '/' ) ? dirname( $plugin_file ) : $plugin_file;
		return in_array( $slug, self::$dependency_slugs, true );
```

costdev commented on src/wp-includes/class-wp-plugin-dependencies.php:
```diff
@@ -0,0 +1,884 @@
+			self::$plugins = get_plugins();
+		}
+
+		$slug = str_contains( $plugin_file, '/' ) ? basename( dirname( $plugin_file ) ) : $plugin_file;
```
```suggestion
		$slug = str_contains( $plugin_file, '/' ) ? dirname( $plugin_file ) : $plugin_file;
```
No need to `basename()`. Everything is stated to be relative to the plugins directory.

costdev commented on src/wp-includes/class-wp-plugin-dependencies.php:
```diff
@@ -0,0 +1,884 @@
+			);
+		} else {
+			// More dependencies to install.
+			$installed_slugs = array_map( 'dirname', array_keys( self::$plugins ) );
```
Need to handle single file plugins here.
```suggestion
			$installed_slugs = array();
			foreach ( array_keys( self::$plugins ) as $plugin ) {
				$installed_slugs[] = str_contains( $plugin, '/' ) ? dirname( $plugin ) : $plugin;
			}
```

costdev commented on src/wp-includes/class-wp-plugin-dependencies.php:
```diff
@@ -0,0 +1,884 @@
+			self::$dependency_slugs        = array_merge( self::$dependency_slugs, $dependency_slugs );
+		}
+
+		$dependent_keys         = array_keys( self::$dependencies );
```
Let's add protection for single file plugins.
```suggestion
		$dependent_keys = array();
		foreach ( array_keys( self::$dependencies ) as $dependency ) {
			$dependent_keys[] = str_contains( $dependency , '/' ) ? dirname( $dependency ) : $dependency;
		}
```

costdev commented on src/wp-includes/class-wp-plugin-dependencies.php:
```diff
@@ -0,0 +1,884 @@
+	 */
+	protected static function get_active_dependents_in_dependency_tree( $plugin_file ) {
+		$all_dependents = array();
+		$slug           = str_contains( $plugin_file, '/' ) ? basename( dirname( $plugin_file ) ) : $plugin_file;
```
No need to use `basename()` here. Everything is stated to be relative to the plugins directory.
```suggestion
		$slug           = str_contains( $plugin_file, '/' ) ? dirname( $plugin_file ) : $plugin_file;
```

costdev commented on src/wp-includes/class-wp-plugin-dependencies.php:
```diff
@@ -0,0 +1,884 @@
+		$all_dependents = array();
+		$slug           = str_contains( $plugin_file, '/' ) ? basename( dirname( $plugin_file ) ) : $plugin_file;
+
+		// Single file plugin.
+		if ( '.' === $slug ) {
+			$slug = basename( $plugin_file );
+		}
+
```
This shouldn't be necessary anymore.
```suggestion
```

afragen committed:
Update class-wp-plugin-dependencies.php

Co-authored-by: Colin Stewart <79332690+costdev@users.noreply.github.com>

costdev commented on src/wp-includes/class-wp-plugin-dependencies.php:
```diff
@@ -0,0 +1,884 @@
+			self::$plugin_dirnames_cache = self::$plugins;
+
+			foreach ( array_keys( self::$plugins ) as $plugin ) {
+				$dirname = dirname( $plugin );
+
+				if ( '.' !== $dirname ) {
+					self::$plugin_dirnames[ $dirname ] = $plugin;
+				} else {
+					// Single file plugin.
+					self::$plugin_dirnames[ $plugin ] = $plugin;
+				}
```
We can simplify the logic with some slugifying.
```suggestion
				$slug = str_contains( $plugin, '/' ) ? dirname( $plugin ) : $plugin;
				self::$plugin_dirnames[ $slug ] = $plugin;
```

afragen committed:
Update class-wp-plugin-dependencies.php

Co-authored-by: Colin Stewart <79332690+costdev@users.noreply.github.com>

afragen committed:
Update class-wp-plugin-dependencies.php

Co-authored-by: Colin Stewart <79332690+costdev@users.noreply.github.com>

afragen committed:
Update class-wp-plugin-dependencies.php

Co-authored-by: Colin Stewart <79332690+costdev@users.noreply.github.com>

costdev commented on src/wp-includes/class-wp-plugin-dependencies.php:
```diff
@@ -0,0 +1,884 @@
+			 *
+			 * Convert $dependent to slug format for checking.
+			 */
+			$dependent_slug = dirname( $dependent );
```
Let's add some protection for single file plugins.
```suggestion
			$dependent_slug = str_contains( $dependent, '/' ) ? dirname( $dependent ) : $dependent;
```

afragen committed:
Update class-wp-plugin-dependencies.php

Co-authored-by: Colin Stewart <79332690+costdev@users.noreply.github.com>

afragen committed:
Update class-wp-plugin-dependencies.php

Co-authored-by: Colin Stewart <79332690+costdev@users.noreply.github.com>

afragen committed:
Update class-wp-plugin-dependencies.php

Co-authored-by: Colin Stewart <79332690+costdev@users.noreply.github.com>

afragen committed:
Update src/wp-includes/class-wp-plugin-dependencies.php

Co-authored-by: Colin Stewart <79332690+costdev@users.noreply.github.com>

afragen committed:
lint

Commit:
Simplify logic to build dependent slugs.

Commit:
Add `get_dependents()` to the public API.

Commit:
Add `get_dependencies()` to the public API.

Commit:
Introduce and implement `convert_to_slug()`.

Commit:
Remove `protected` `get_dependents()` and `get_dependencies()`.

Commit:
Guard against missing filepath in `get_dependency_filepath()`.

Commit:
Remove unnecessary `self::$plugins` guard.

Produces an untestable branch later in `get_dependency_filepaths()`.
`get_plugin_dirnames()` has a check which will limit processing in lieu
of this guard.

Commit:
Introduce `get_dependency_filepath()`.

Commit:
Fix erroneous use of `(string) 'requires_plugins'` as an array.

afragen committed:
Merge pull request #5 from costdev/plugin-dependencies-updates

Updates to Plugin Dependencies

Commit:
Update tests after refactor.

afragen committed:
update ::get_dependencies_notice()

afragen committed:
Merge branch 'trunk' into dependencies-tab

afragen committed:
lint

afragen committed:
missing variable in sprintf

afragen committed:
update docblock

afragen committed:
oops 🐿️

afragen committed:
Merge branch 'trunk' into dependencies-tab

afragen committed:
lint

afragen committed:
update since

afragen committed:
not needed

afragen committed:
split notices into separate functions

afragen committed:
too many %s, bad copy/paste

afragen committed:
remove outdated tests

afragen committed:
Merge branch 'trunk' into dependencies-tab

afragen committed:
remove outdated variable

afragen committed:
linting

afragen committed:
more outdated variables

azaozz committed:
Merge remote-tracking branch 'upstream/trunk' into pr/3032

Commit:
Bootstrap/Load: Don't load WP_Plugin_Dependencies during test runs.

Commit:
Require `WP_Plugin_Dependencies` in its tests.

afragen committed:
Merge pull request #7 from costdev/plugin-dependencies/do_not_load_in_bootstrap_during_test_suite

Don't load in bootstrap during test suite.

afragen committed:
Merge pull request #8 from costdev/plugin-dependencies-test-updates

Test updates.

costdev commented on tests/phpunit/tests/admin/plugin-dependencies/getDependents.php:
```diff
@@ -0,0 +1,45 @@
+			array(
+				'dependent/dependent.php'   => array( 'dependency' ),
+				'dependent2/dependent2.php' => array( 'dependency' ),
+			),
```
```suggestion
			)
```

costdev committed:
Remove errant comma in tests.

azaozz committed:
Merge branch 'dependencies-tab' of https://github.com/afragen/wordpress-develop into pr/3032

azaozz committed:
Merge remote-tracking branch 'upstream/trunk' into pr/3032

afragen committed:
Merge branch 'trunk' into dependencies-tab

azaozz committed:
Merge branch 'dependencies-tab' of https://github.com/afragen/wordpress-develop into pr/3032

Commit:
Database: Store plugin data from `get_plugins()`.

Commit:
Database: Remove plugin data for a deleted plugin.

Commit:
Bootstrap/Load: Check active plugins for missing requirements.

Commit:
Upgrade/Install: Set/Update a plugin's data on package install.

Commit:
Plugin Dependencies: Load from the database.

Commit:
Plugin Dependencies: Require `wp-admin/includes/plugin.php` at the start.

Functions from that file are used in this class.

Commit:
Plugin Dependencies: Make `self::get_plugins()` use the database first, then the filesystem.

Commit:
Plugin Dependencies: Ensure plugin data exists when getting dirnames.

Commit:
Plugin Dependencies: Use `self::get_plugins()` getting dependent names.

Commit:
Plugin Dependencies: Accommodate Hello Dolly when converting to a slug.

In Core's `get_plugins()`, `hello.php` is converted to `hello-dolly`.

Previously, `self::convert_to_slug()` converted this to `hello`.

This adds a special case for Hello Dolly to ensure it gets the correct slug.

Commit:
Bootstrap/Load: Only show notices on Dashboard and `plugins.php`.

Commit:
Bootstrap/Load: Add a check for unmet dependencies.

Commit:
Plugin Dependencies: Only load `wp-admin/includes/plugin.php` where needed.

azaozz committed:
Fix missing action links, add notices

azaozz committed:
Umm, remove blank line despite making it harder to read...

azaozz committed:
Fix white space in src/js/_enqueues/wp/updates.js

Co-authored-by: Mukesh Panchal <mukeshpanchal27@users.noreply.github.com>

azaozz committed:
Use printf() to output dependency warnings in src/wp-admin/includes/class-wp-plugins-list-table.php

Co-authored-by: Colin Stewart <79332690+costdev@users.noreply.github.com>

azaozz committed:
Fix the 'since', and improve the warnings text

Commit:
Updates: Do not close the modal on click.

Commit:
Updates: Add logic for activating a plugin via AJAX.

Commit:
Updates: Remove modal closure on install/update.

Commit:
Updates: Add the modal's footer to selectors.

Commit:
Updates: Abstract dependencies check logic to methods.

Commit:
Updates: Add the dependencies check to `isValidResponse()`.

Commit:
AJAX: Add `wp_ajax_activate_plugin()`.

Commit:
AJAX: Add `activate-plugin` to `$core_actions_post`.

Commit:
Bootstrap/Load: Add `activate-plugin` to the protected actions.

Commit:
Plugins: Add some extra data to `wp_get_plugin_action_button()`.

Commit:
Plugin Dependencies: Add some extra data to `::check_plugin_dependencies_during_ajax()`.

Commit:
Updates: Add/protect stying for the buttons in cards and modals.

Commit:
Plugins: Print the filesystem credentials modal in the plugin information modal.

Modalception.

Commit:
Updates: Ensure the Update Now button uses the proper selectors.

Also skips decrementing the plugin count when inside the modal.

Commit:
Updates: Add some styling for the Update Now button in the modal.

Commit:
Updates: Improve accessibility.

Commit:
Plugins: Print the admin notice templates in the plugin information modal.

Commit:
Updates: Add some styling for the notices in the modal.

Commit:
Updates: When a modal button updates, also update the card button.

afragen committed:
Merge pull request #11 from azaozz/Plugin-Dependencies-add-notices-and-warnings

Plugin dependencies: add/fix notices and warnings

afragen committed:
Merge branch 'plugin_dependencies_database_integration' into dependencies-tab

afragen committed:
fix for empty database

afragen committed:
linting

Commit:
Bootstrap/Load: Remove invalid plugins from `plugin_data` option.

Commit:
Plugins: Prevent plugin activation for unmet plugin dependencies.

Commit:
Plugin Dependencies: Only get dependency API data on `plugins.php` and `plugin-install.php`.

afragen committed:
Merge pull request #13 from costdev/remove_invalid_plugins_from_plugin_data

Plugin Dependencies: Remove invalid plugins from `plugin_data` option

afragen committed:
Merge pull request #15 from costdev/only_get_dependency_api_data_on_plugins_and_plugin-install

afragen committed:
Merge pull request #14 from costdev/prevent_activate_plugin_from_activating_plugins_with_unmet_dependencies

Plugin Dependencies: Prevent `activate_plugin()` from activating plugins with unmet dependencies.

Commit:
Bootstrap/Load: Only make `WP_Plugin_Dependencies::initialize()` conditional.

Previously, the inclusion of the `WP_Plugin_Dependencies()` was also conditional,
causing a Fatal Error during test runs with internal callstacks requiring it.

Commit:
Tests: Remove unnecessary `require_once()`.

Commit:
Tests: Store reflected members rather than create new ones.

Commit:
Tests: Add a test to make sure `hello.php` is converted to `hello-dolly`.

afragen committed:
Merge pull request #16 from costdev/require_pd_but_just_dont_initialize_during_tests

Plugin Dependencies: Require the class in Core's bootstrap, but just don't initialize it during tests.

afragen committed:
Merge pull request #17 from costdev/pd_tests_store_reflected_members_instead_of_creating_new_ones

Plugin Dependencies: Store reflected members instead of creating new ones in tests.

afragen committed:
Merge pull request #18 from costdev/pd_tests_add_hello_dolly_test_for_convert_to_slug

Plugin Dependencies: Add a test to ensure that `hello.php` is converted to `hello-dolly`.

afragen committed:
Merge branch 'WordPress:trunk' into dependencies-tab

afragen committed:
Merge branch 'dependencies-tab' into plugin_dependencies_modal_handling

afragen committed:
Merge pull request #12 from costdev/plugin_dependencies_modal_handling

Plugin Dependencies: Make modals persistent for Install Now, Activate and Update Now.

afragen committed:
something different here

Commit:
Coding Standards: Fix issues reported by JSHint.

afragen committed:
Merge pull request #19 from costdev/pd_jshint

Plugin Dependencies: Fix issues reported by JSHint.

Commit:
Source changes.

Commit:
Test changes

costdev committed:
Plugins: Correct key for `plugin_data` in `get_plugins()`.

afragen committed:
Merge pull request #20 from costdev/pd_fix_get_plugins_update_of_plugin_data

Plugin Dependencies: Fix the key `get_plugins()` sets when updating plugin data.

afragen committed:
Merge pull request #21 from costdev/pd_tests_update

Plugin Dependencies: Tests update

afragen committed:
linting

costdev committed:
Fix detection for circular dependencies.

costdev committed:
Show 'Delete' for circular dependencies.

costdev committed:
Update tests.

costdev committed:
Update circular dependency messaging.

afragen committed:
Merge pull request #22 from costdev/pd_fix_circular_dependencies

Plugin Dependencies: Fix circular dependencies

costdev committed:
Add missing `::initialize()` call

afragen committed:
errant comma

afragen committed:
update since

afragen committed:
change scope to protected

afragen committed:
add @ticket to tests

costdev commented:
Thanks for the PR! Committed in [r57545](https://core.trac.wordpress.org/changeset/57545).
